src/main/java/javaslang/collection/Stream.java

```diff
@@ -111,6 +111,21 @@
     }
 
     /**
+     * Builds an (theoretically) infinitely long Stream using a function to calculate the tail of the stream
+     * based on the head.
+     *
+     * @param head         The first value in the Stream
+     * @param tailFunction A function to calculate the tail values based on the head. To end the stream,
+     *                     return {@link Stream#nil}.
+     * @param <T>          value type
+     * @return A new Stream
+     */
+    static <T> Stream<T> build(T head, Function<T, Stream<T>> tailFunction) {
+        Objects.requireNonNull(tailFunction, "tailFunction is null");
+        return new Stream.Cons<>(head, () -> tailFunction.apply(head));
```

Comment:
We apply the head that is already given, so we can apply it before the call of build. I will give an example in a comment below...

Then we can rename `build` to `gen` and use `Supplier<Stream<T>>` instead of the `Function`.